### PR TITLE
Expose guest topology to tests and scripts

### PR DESCRIPTION
**Pull request by happz**
`multihost` plugin took care of setting some multihost-related envvars,
plus it added entries to `/etc/host`. Unfortunately, being part of
`prepare` step, these changes would not happen when the `prepare` step
was skipped, or in single-host scenarios.

The functionality is now part of the core step implementation, and
therefore applied to all guest-backed steps (`prepare`, `execute`,
`finish`). It no longer makes modifications to `/etc/hosts`, instead tmt
exposes guest-related information in files whose path is stored in
well-defined environment variables.

Tests - and prepare or finish scripts - may then import these files, and
use provided information to reach every other guest. Guest names are not
resolvable, instead a script may acquire the hostname/IP from the
topology file.

This drops the need for environment variables like `TMT_GUESTS`,
`TMT_ROLES` or `TMT_GUEST_*` - all necessary information is now saved in
files.

### CODE REVIEW

File: tmt/steps/provision/__init__.py [hunk 1]
```diff
@@ -1486,6 +1486,46 @@ def save(self) -> None:
         except tmt.utils.FileError:
             self.debug('Failed to save provisioned guests.')
 
+    def prepare_environment(self, guests: List[Guest]) -> None:
+        # Spike plan environment with multihost-related variables.
```
Comment (lukaszachy) [inline tmt/steps/provision/__init__.py:1490]: Could this be a docstring instead?

File: tmt/steps/provision/__init__.py [hunk 2]
```diff
@@ -1486,6 +1486,46 @@ def save(self) -> None:
...
+
+        # A list of all roles in `TMT_ROLES`.
+        self.plan._environment['TMT_ROLES'] = ' '.join(roles.keys())
+
+        # For each role, `TMT_ROLE_$rolename` lists all guests with this role.
+        for role, guests in roles.items():
+            self.plan._environment[f'TMT_ROLE_{role}'] = ' '.join(guest.name for guest in guests)
+
+    def prepare_resolving(self, guests: List[Guest]) -> None:
+        # Add entries for guests to /etc/hosts
```
Comment (lukaszachy) [inline tmt/steps/provision/__init__.py:1513]: Could this be a docstring instead?

File: tmt/steps/provision/__init__.py (hunk 1 above)
Comment (happz) [reply tmt/steps/provision/__init__.py]: Fixed in 9ad274fa

**Comment by happz** [comment]
> tests/full finished and /tests/execute/nonroot is failing.
> 
> The /etc/hosts manipulation means that prepare shell has to run as root, we techniclly support other users as long as they have configured 'sudo' correctly.
> 
> I'm not sure how to solve that, we are missing way how to make tmt aware that prepare -h shell script needs privileged user. Personally I'd be against running all scripts as root (or via sudo).

Ouch. Yeah, not going to run everything with root.

On the other hand, this is a prepare script, but something tmt runs. Don’t we have some “has sudo” flag somewhere we could use, and actually obey here? This might be the actual issue, patch not using “use sudo” flag correctly.

**Comment by happz** [comment]
> > On the other hand, this is a prepare script, but something tmt runs. Don’t we have some “has sudo” flag somewhere we could use, and actually obey here? This might be the actual issue, patch not using “use sudo” flag correctly.
> 
> Hm, something similar was done in [a83e73b](https://github.com/teemtee/tmt/commit/a83e73b69a346735fe30424b7b27ba50c48e9b5c) but it might be worth to have it as a 'fact' now

That's it! I couldn't find it :/ Ok, I'll try to amend these two into some solution, maybe we'll need an extra patch to introduce the fact. Should be quick to resolve.

**Comment by happz** [comment]
> > > On the other hand, this is a prepare script, but something tmt runs. Don’t we have some “has sudo” flag somewhere we could use, and actually obey here? This might be the actual issue, patch not using “use sudo” flag correctly.
> > 
> > 
> > Hm, something similar was done in [a83e73b](https://github.com/teemtee/tmt/commit/a83e73b69a346735fe30424b7b27ba50c48e9b5c) but it might be worth to have it as a 'fact' now
> 
> That's it! I couldn't find it :/ Ok, I'll try to amend these two into some solution, maybe we'll need an extra patch to introduce the fact. Should be quick to resolve.

Another shot at `/etc/hosts` in https://github.com/teemtee/tmt/pull/2072/commits/aa5a0dbb155bd4f06cf923e682d2ff435ee8bfcf, this time with added `sudo` when needed.

**Comment by kkaarreell** [comment]
There is something not working for me with this PR in place. I do not know details yet but I suspect that with this change tmt is not passing a `guest` in the `TMT_ROLE_xxx` when `provision: connect` is used and instead it uses the name of the role. Please reach out to me directly next week eventually, I can help you to do the setup or prepare a one for troubleshooting.

File: tmt/steps/prepare/multihost.py [hunk 3]
```diff
@@ -1,67 +0,0 @@
...
-        # Find out which one.
-        # See https://github.com/teemtee/tmt/pull/2072
-#        logger.debug('Export roles.', level=2)
-#        for role, corresponding_guests in self.get('roles').items():
-#            self.step.plan._environment[f'TMT_ROLE_{role}'] = ' '.join(corresponding_guests)
-        logger.debug("Add hosts to '/etc/hosts'.", level=2)
-        for host_name, host_address in self.get('hosts').items():
-            if host_address:
-                guest.execute(
-                    ShellScript(f'echo "{host_address} {host_name}" >> /etc/hosts'),
```
Comment (thrix) [inline tmt/steps/prepare/multihost.py:66]: I am blind, where is `/etc/hosts` in the new code populated?

**Review by thrix — CHANGES_REQUESTED** [review]

**Comment by thrix** [comment]
@kkaarreell would be nice if you could test this one out ....

**Comment by happz** [comment]
Rebased & the main commit message was amended heavily to reflect the changes.

File: README.rst [hunk 4]
```diff
@@ -428,25 +428,90 @@ TMT_REBOOT_COUNT
...
-    contains a list of all guests on which the test is running at
-    the moment.
-
-    This variable will be removed in the future, it is supported only
-    to make multihost prototype support easier to verify as many current
-    tests were developed with Restraint.
+TMT_TOPOLOGY_SH, TMT_TOPOLOGY_YAML
+    Paths of files describing existing guest, their roles and the
+    guest on which the test is running. Both files contain the very
+    same information, ``TMT_TOPOLOGY_SH`` can be consumed by Bash 4
```
[the remark this answers is not in the document]
Comment (happz) [reply README.rst]: Good question, I'm in favor of `SH`, but I don't mind, no preference.

**Comment by The-Mule** [comment]
I tested this PR running the following test plan for https://github.com/beakerlib/sync/tree/master/sync runtest.sh:

```
summary: sync library self-test plan

provision:
  - name: server
    how: connect
    guest: <hostname1>
    user: root
    role: server
  - name: client
    how: connect
    guest: <hostname2>
    user: root
    role: client

discover:
  - name: self-test
    how: fmf
    where:
      - server
      - client
    test:
      - /sync

execute:
  how: tmt

prepare:
    # To get /distribution/sync-internal we need beaker-tasks library.
    - name: Add beaker-tasks repository
      where:
        - server
        - client
      how: shell
      script:
        - dnf config-manager --enable beaker-tasks
```

`$ tmt run -a -vvv plans --name testplan report -vvv finish`

And it worked fine for me. Since I am expecting 'server' and 'client' roles I used the new shell variables in my library as follows:

```
    if [ -f "$TMT_TOPOLOGY_SH" ]; then
        source $TMT_TOPOLOGY_SH
        export SERVERS=${TMT_GUESTS["server.hostname"]}
        export CLIENTS=${TMT_GUESTS["client.hostname"]}
    fi
```

**Comment by kkaarreell** [comment]
I have just tested this PR on my multi-host test and it worked well. The only issue I have noticed is that `TMT_TOPOLOGY_SH` is not exported when `tmt run ... login` but that is not serious and I think it would deserve separate PR.

File: README.rst (hunk 4 above)
[the remark this answers is not in the document]
Comment (psss) [reply README.rst]: I'd agree with @thrix, if the `bash` arrays are used and the functionality is thus limited to `bash`, let's call it `TMT_TOPOLOGY_BASH`.

File: README.rst [hunk 5]
```diff
@@ -428,25 +428,10 @@ TMT_REBOOT_COUNT
...
-SERVERS
-    A legacy variable, exported for the backward compatibility. It
-    contains a list of all guests on which the test is running at
-    the moment.
-
-    This variable will be removed in the future, it is supported only
-    to make multihost prototype support easier to verify as many current
-    tests were developed with Restraint.
+TMT_TOPOLOGY_SH, TMT_TOPOLOGY_YAML
+    Paths of files describing existing guest, their roles and the
```
[the remark this answers is not in the document]
Comment (happz) [reply README.rst]: Fixed in 77e19aff

File: README.rst (hunk 4 above)
[the remark this answers is not in the document]
Comment (happz) [reply README.rst]: Fixed in 77e19aff

**Comment by psss** [comment]
This change breaks `/tests/multihost/web` test as it relies on the server hostname to be resolvable:

    Resolving httpd-server (httpd-server)... failed: Name or service not known.

It's run as part of the full test suite only.

File: README.rst (hunk 4 above)
[the remark this answers is not in the document]
Comment (kkaarreell) [reply README.rst]: > I'd agree with @thrix, if the `bash` arrays are used and the functionality is thus limited to `bash`, let's call it `TMT_TOPOLOGY_BASH`.

I do not. For bash scripts we are not using bash suffix either. Also, this array definition should be compatible also with zsh.

**Comment by happz** [comment]
> This change breaks `/tests/multihost/web` test as it relies on the server hostname to be resolvable:
> 
> ```
> Resolving httpd-server (httpd-server)... failed: Name or service not known.
> ```
> 
> It's run as part of the full test suite only.

Should be fixed with the following patch. I will try to verify it tomorrow.

```diff
diff --git a/tests/multihost/web/data/test.fmf b/tests/multihost/web/data/test.fmf
index ff6efb31..902f57a7 100644
--- a/tests/multihost/web/data/test.fmf
+++ b/tests/multihost/web/data/test.fmf
@@ -11,11 +11,13 @@
 /client:
     require: [curl, wget]
     test: |
+        . "$TMT_TOPOLOGY_BASH"
+
         [[ $(hostname) == curl-client ]] && command="curl --silent"
         [[ $(hostname) == wget-client ]] && command="wget --quiet -O -"
 
         for count in {1..33}; do
             echo "Get file using '$command', attempt $count."
-            $command http://httpd-server/ | grep -q foo
+            $command "http://${TMT_GUESTS[httpd-server.hostname]}/" | grep -q foo
             sleep .$[$RANDOM % 10]
         done
```

**Comment by psss** [comment]
Thanks for the fix. Verified and committed as a8a346f0.

File: spec/plans/guest-topology.fmf [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+        :caption: TMT_TOPOLOGY_YAML
+
+        # Guest on which the test or script is running.
+        guest:
+            name: ...
+            role: ...
+            hostname: ...
+
+        # List of names of all provisioned guests.
+        guest_names:
```
Comment (psss) [inline spec/plans/guest-topology.fmf:39]: So far we're consistently using `-` for separating multiword keys in `yaml` files (except a few exceptions which slipped through :P ) Could we continue in that direction here as well?

File: spec/plans/guest-topology.fmf [hunk 7]
```diff
@@ -0,0 +1,114 @@
+summary: Define format of on-disk description of provisioned guest topology
+title: Guest topology format
```
Comment (psss) [inline spec/plans/guest-topology.fmf:2]:
```suggestion
title: Guest Topology Format
```

Consistent with other headings in the section.

File: tests/multihost/complete/data/plans.fmf [hunk 8]
```diff
@@ -21,12 +21,16 @@ prepare:
   - how: shell
     script:
       - for envvar in $(compgen -A variable | egrep 'SERVERS|TMT' | sort); do echo "$envvar=${!envvar}"; done
```
Comment (psss) [inline tests/multihost/complete/data/plans.fmf:23]: The `SERVERS` variable has been removed?

File: tests/multihost/prepare/main.fmf [hunk 9]
```diff
@@ -1,2 +0,0 @@
-summary: Verify that multihost preparation works
-duration: 10m
```
Comment (psss) [inline tests/multihost/prepare/main.fmf:2]: Coverage from this test has been moved to `/tests/multihost/complete`?

**Review by psss — CHANGES_REQUESTED** [review]
Looks nice. Added a couple of comments.

File: tests/multihost/prepare/main.fmf (hunk 9 above)
Comment (happz) [reply tests/multihost/prepare/main.fmf:2]: Correct, `/tests/multihost/complete` checks `prepare` step as well.

Comment (psss) [reply tests/multihost/prepare/main.fmf:2]: Good, thanks for clarification.

File: spec/plans/guest-topology.fmf (hunk 6 above)
Comment (happz) [reply spec/plans/guest-topology.fmf:39]: Fixed in 2526d401.

File: spec/plans/guest-topology.fmf (hunk 7 above)
Comment (happz) [reply spec/plans/guest-topology.fmf]: Fixed in 2526d401.

File: tests/multihost/complete/data/plans.fmf (hunk 8 above)
Comment (happz) [reply tests/multihost/complete/data/plans.fmf]: Correct, it's been removed. Test no longer looks for it after 2526d401.

File: spec/plans/guest-topology.fmf (hunk 6 above)
Comment (psss) [reply spec/plans/guest-topology.fmf:39]: Thanks. The spec example still uses `guest_names` and `role_name` though.

Comment (happz) [reply spec/plans/guest-topology.fmf]: Fixed in 9950ce74.

**Review by psss — APPROVED** [review]

**Pull request merged**